### DISCUSSION TITLE
linux: wrong mouse cursor size on multi-monitor setup

### DISCUSSION THREAD

**Discussion by HE7086** [ghostty-org/ghostty#3167] [Q&A] [answered]
I have a kind of cursed display setup:
Display 1: 3840x2160 @ 150%
Display 2: 3840x2160 @ 200%
Display 3: 1920x1080 @ 100%

Ghostty is handling the cursor size incorrectly on 1 and 2.
The recording attached is at the connecting edge of 1(left) and 3(right).
All other applications are working properly.
I have tried to set all 3 displays to 100% scale and the problem is resolved.

System information:
WM: Kwin wayland (KDE Plasma 6.2.4)
Cursor: Breeze_Light (36px)
GPU: NVIDIA GeForce RTX 3080 Ti

Ghostty 1.0.0-main+4b4d4062 (built from AUR ghostty-git)
```
Version
  - version: 1.0.0-main+4b4d4062
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.7
    runtime    : 4.16.7
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
```
https://github.com/user-attachments/assets/5ac4d7a0-c0d6-451a-a308-b99b8f46e7ca

**Comment by mitchellh** [comment]
Converting to Q&A because I recall some users in the beta hitting this and it's a desktop config issue. We don't render or configure the cursor in any way.

**Comment by HE7086** [comment, marked as answer]
Looks like a gtk issue:
https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7760
https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/6212
https://blogs.kde.org/2024/10/09/cursor-size-problems-in-wayland-explained/

should be fixed in next gtk release (?)

---

edit: confirm fixed in gtk 4.18

**Comment by pbek** [comment]
I have the same issue on KDE Plasma 6 under Wayland. So far I haven't seen that issue with other applications, does anyone know a workaround?

```
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.3
    runtime    : 4.16.3
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
```

**Reply by HE7086** [reply]
Atm you could compile ghostty against other gtk version to workaround that. Any version other than 4.16 should work (read the kde blog mentioned above). But then your other gtk apps might stop working.

**Reply by pbek** [reply]
Wouldn't the runtime still have the issue? 🤔

**Reply by HE7086** [reply]
Yeah I meant downgrading the system gtk runtime (or install directly from git source), which might cause issues with other apps.

**Comment by topazus** [comment]
I also encountered the same cursor issue with gtk 4.17.1.
```bash
❯ ghostty --version
Ghostty 1.0.0-dev+0000000

Version
  - version: 1.0.0-dev+0000000
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.17.1
    runtime    : 4.17.1
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
```

**Reply by tg90nor** [reply]
Interesting. I assumed https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7760 would fix this, and that is included in 4.17.0 and above.

**Reply by tg90nor** [reply]
Do you use the breeze cursor theme?

**Comment by xulongwu4** [comment]
Setting the `XCURSOR_THEME` env variable to my cursor theme fixed it for me.

**Reply by pbek** [reply]
Hm, didn't help for me. Neither did `XCURSOR_SIZE`... What did you set?

**Reply by xulongwu4** [reply]
I guess it depends on what cursor theme you use now. I set it to `Nordic-cursors`. Then a reboot fixed it.

**Reply by xulongwu4** [reply]
My issue was more about the cursor style than the cursor size, so your problem might be different.

**Reply by tg90nor** [reply]
That's not what this issue is about. It's specifically about the cursor size scaling with the breeze cursor theme (and other themes with bigger image size than nominal size)

**Reply by tg90nor** [reply]
I'm glad you got your issue fixed though :slightly_smiling_face:

**Comment by ex1tium** [comment]
Launching on Endeavour OS KDE Plasma 6 (Wayland) with `GDK_SCALE=1` seems to fix this:

`GDK_SCALE=1 ghostty`

or not... if the display is in portrait mode the issue comes back, seems to work on landscape mode displays. I have two portrait and two landscape monitors. Scaling set to 100% on each in Display Configuration.

**Reply by pbek** [reply]
Hm, also doesn't help on NixOS with Plasma 6 (Wayland). 🤔

**Reply by nduartech** [reply]
With nwg-look switching my cursor theme to Qogir worked

**Comment by iynaix** [comment]
Fixed for me (Hyprland on NixOS):

```nix
(pkgs.ghostty.override {
    wrapGAppsHook4 = pkgs.wrapGAppsNoGuiHook.override {
      isGraphical = true;
      gtk3 =
        pkgs.__splicedPackages.gtk4.overrideAttrs
          (o: rec {
            version = "4.17.6";
            src = pkgs.fetchurl {
              url = "mirror://gnome/sources/gtk/${lib.versions.majorMinor version}/gtk-${version}.tar.xz";
              hash = "sha256-366boSY/hK+oOklNsu0UxzksZ4QLZzC/om63n94eE6E=";
            };
            patches = [ ]; # asahi patch doesn't apply
            postFixup = ''
              demos=(gtk4-demo gtk4-demo-application gtk4-widget-factory)

              for program in ''${demos[@]}; do
                wrapProgram $dev/bin/$program \
                  --prefix XDG_DATA_DIRS : "$GSETTINGS_SCHEMAS_PATH:$out/share/gsettings-schemas/${o.pname}-${version}"
              done

              # Cannot be in postInstall, otherwise _multioutDocs hook in preFixup will move right back.
              moveToOutput "share/doc" "$devdoc"
            '';
          });
    };
  })
  ```
  
  Edit: wayland-protocols has been updated to 1.41, that override is no longer necessary

**Reply by pbek** [reply]
Worked for me on NixOS Plasma 6 Wayland multiscreen!

**Reply by bizmythy** [reply]
incredible, exactly what i needed, worked for me on nixos hyprland (with phinger-cursors-light)

**Reply by bizmythy** [reply]
can confirm, issue now resolved on latest hyprland

**Comment by Quanalogy** [comment]
I just wanted to chime in on this. I experience this issue as well on Fedora 41, KDE.
I have a multimonitor setup where one screen is in portrait mode and I only have the issue on that screen.

relevant version info:

```bash
$ ghostty --version
Ghostty 1.1.2-1.fc41

Version
  - version: 1.1.2-1.fc41
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - desktop env: other
  - GTK version:
    build      : 4.16.5
    runtime    : 4.16.5
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
  - libX11     : enabled
  - libwayland : enabled

$ plasmashell --version
plasmashell 6.3.3
```

**Comment by Renari** [comment]
Ran into this today, tried a few things in this thread and nothing worked, but my situation is a bit odd it only happens on my vertical monitor (which is set to 100% scale so it's not scaling related).

```
(process:179737): Gdk-WARNING **: 17:34:59.389: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (32) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:01.060: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (32) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:01.062: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:08.222: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:09.116: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:11.080: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:11.980: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
(process:179737): Gdk-WARNING **: 17:35:28.053: ../gtk/gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
```
Gdk seems to be doubling the size of the cursor when it's on my vertical monitor for some reason.

**Reply by Renari** [reply]
Seems this is a GTK bug that is due to be fixed in 4.18:
https://bbs.archlinux.org/viewtopic.php?id=299624
